spynnaker/pyNN/models/neuron/population_vertex.py

```diff
@@ -474,35 +498,195 @@ def set_structure(self, structure: BaseStructure) -> None:
+            spikes_per_second = self.__spikes_per_second
+            if isinstance(pre_vertex, AbstractMaxSpikes):
+                rate = pre_vertex.max_spikes_per_second()
+                if rate > 0:
+                    spikes_per_second = rate
+
+            pre_synapses_per_second: int = 0
+            for proj in projs:
+                # pylint: disable=protected-access
+                s_info = proj._synapse_information
+                dynamics = s_info.synapse_dynamics
+                conn = s_info.connector
+                n_conns: Optional[int] = None
+                if isinstance(dynamics, AbstractSDRAMSynapseDynamics):
+                    n_conns = dynamics.pad_to_length
+                if n_conns is None:
+                    n_conns = conn.get_n_connections_from_pre_vertex_maximum(
+                        self.get_max_atoms_per_core(), s_info)
+                # The number of synapses is the number of connections from each
+                # pre-neuron to each post-neuron
+                assert n_conns is not None
+                pre_synapses_per_second += math.ceil(
+                    n_conns * spikes_per_second * pre_vertex.n_atoms)
+
+            if self._is_direct_poisson(pre_vertex, projs):
+                poisson_synapses_per_second += pre_synapses_per_second
+            else:
+                synapses_per_second += pre_synapses_per_second
+
+        # How many cores are needed to process the non-direct-poisson synapses?
+        n_synapse_cores = math.ceil(
+            synapses_per_second / synapses_per_core_per_sim_second)
+
+        # The number of Poisson cores that will be needed
+        n_poisson_cores = len(self.incoming_poisson_projections)
+
+        # If we can definitely do the Poisson vertices directly,
+        # lets recommend this
+        if n_synapse_cores + n_poisson_cores < max_n_cores:
+            self.__n_synapse_cores = n_synapse_cores
+        else:
+
+            # Otherwise, we should consider how many more cores we need for the
+            # Poisson input spikes
+            n_synapse_cores = math.ceil(
+                (synapses_per_second + poisson_synapses_per_second) /
+                synapses_per_core_per_sim_second)
+
+            # If the number of cores needed is more than the maximum, use the
+            # maximum
+            self.__n_synapse_cores = min(max_n_cores, n_synapse_cores)
+
+    @property
+    def max_delay_steps(self) -> int:
+        """
+        The maximum number of delay steps supported on a core.
+        """
+        self.__update_max_delay()
+        assert self.__max_delay_slots_available is not None
+        return self.__max_delay_slots_available
+
+    @property
+    def max_delay_steps_incoming(self) -> int:
+        """
+        The maximum delay steps needed to handle incoming synapses,
+        accounting for delay extensions.
+        """
+        self.__update_max_delay()
+        assert self.__max_delay_ms is not None
```

Comment:
max_delay = self.def max_delay_steps

It that method (merged or not) do the update and the assert it is an int

Reply:
I have used the return types from __update_max_delay now